mmap-alloc/src/lib.rs

```diff
@@ -411,90 +277,190 @@ impl MapAlloc {
         if new_layout.align() > self.pagesize {
-            return Err(AllocErr::invalid_input("cannot support alignment greater than a page"));
+            return Err(AllocErr::invalid_input(
+                "cannot support alignment greater than a page",
+            ));
         }
-        debug_assert!(layout.size() > 0 && new_layout.size() > 0,
-                      "usable_size: size of layout and new_layout must be non-zero");
 
         let old_size = next_multiple(layout.size(), self.pagesize);
         let new_size = next_multiple(new_layout.size(), self.pagesize);
         if old_size == new_size {
             return Ok(ptr);
         }
-        self.realloc_helper(ptr, old_size, new_layout.size()).ok_or(AllocErr::Exhausted { request: new_layout })
+        remap(ptr, old_size, new_layout.size(), false).ok_or(AllocErr::Exhausted {
+            request: new_layout,
+        })
+    }
+
+    #[cfg(any(target_os = "macos", windows))]
+    unsafe fn realloc(
+        &mut self,
+        ptr: *mut u8,
+        layout: Layout,
+        new_layout: Layout,
+    ) -> Result<*mut u8, AllocErr> {
+        // Adapted from the Rust standard library.
+
+        debug_assert!(
+            layout.size() > 0 && new_layout.size() > 0,
+            "realloc: size of layout and new_layout must be non-zero"
+        );
+
+        // alignment less than a page is fine because page-aligned objects are also aligned to
+        // any alignment less than a page
+        if new_layout.align() > self.pagesize {
+            return Err(AllocErr::invalid_input(
+                "cannot support alignment greater than a page",
+            ));
+        }
+
+        let old_size = next_multiple(layout.size(), self.pagesize);
+        let new_size = next_multiple(new_layout.size(), self.pagesize);
+
+        if old_size == new_size {
+            return Ok(ptr);
+        }
+
+        if !cfg!(windows) {
+            // Windows cannot grow or shrink existing mappings.
+            if new_size >= old_size {
+                if let Ok(()) = self.grow_in_place(ptr, layout.clone(), new_layout.clone()) {
+                    return Ok(ptr);
+                }
+            } else if new_size < old_size {
+                if let Ok(()) = self.shrink_in_place(ptr, layout.clone(), new_layout.clone()) {
+                    return Ok(ptr);
+                }
+            }
+        }
+
+        // otherwise, fall back on alloc + copy + dealloc.
+        let result = Alloc::alloc(self, new_layout);
+        if let Ok(new_ptr) = result {
+            use core::cmp;
+
+            // Temporarily mark the old allocation readable and the new allocation writable in case
+            // we're configured to allocate un-readable or un-writable memory - we'll need to be
+            // able to read/write in order to copy the data.
+            protect(ptr, old_size, perms::PROT_READ);
+            protect(new_ptr, new_size, perms::PROT_WRITE);
```

Comment:
Even though reallocation should be infrequent, I'm not sure how I feel about 3 guaranteed syscalls, when in most cases none would be needed (I'd expect most allocators to expect read+write)

Reply:
Good call. I'll check the perms first.